codec-dns/src/main/java/io/netty/handler/codec/dns/TcpDnsResponseEncoder.java

```diff
@@ -0,0 +1,103 @@
+/*
+ * Copyright 2021 The Netty Project
+ *
+ * The Netty Project licenses this file to you under the Apache License,
+ * version 2.0 (the "License"); you may not use this file except in compliance
+ * with the License. You may obtain a copy of the License at:
+ *
+ *   https://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+ * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
+ * License for the specific language governing permissions and limitations
+ * under the License.
+ */
+package io.netty.handler.codec.dns;
+
+import io.netty.buffer.ByteBuf;
+import io.netty.channel.ChannelHandler;
+import io.netty.channel.ChannelHandlerContext;
+import io.netty.handler.codec.MessageToMessageEncoder;
+import io.netty.util.internal.ObjectUtil;
+
+import java.util.List;
+
+@ChannelHandler.Sharable
+public final class TcpDnsResponseEncoder extends MessageToMessageEncoder<DnsResponse> {
+    private final DnsRecordEncoder encoder;
+
+    /**
+     * Creates a new encoder with {@linkplain DnsRecordEncoder#DEFAULT the default record encoder}.
+     */
+    public TcpDnsResponseEncoder() {
+        this(DnsRecordEncoder.DEFAULT);
+    }
+
+    /**
+     * Creates a new encoder with the specified {@code encoder}.
+     */
+    public TcpDnsResponseEncoder(DnsRecordEncoder encoder) {
+        this.encoder = ObjectUtil.checkNotNull(encoder, "encoder");
+    }
+
+    @Override
+    protected void encode(ChannelHandlerContext ctx, DnsResponse response, List<Object> out) throws Exception {
+        ByteBuf buf = ctx.alloc().ioBuffer(1024);
```

Comment:
The code mostly looks identical to the UDP based one... Can we extract this to a utility class and share it ?

Reply:
Of course. I try to extract this as shown below, is it ok?

DnsMessageUtil
```
    static DnsQuery decodeDnsQuery(ByteBuf buf, DnsQuerySupplier supplier, DnsRecordDecoder decoder) throws Exception {
        DnsQuery query = newQuery(buf, supplier);
        boolean success = false;
        try {
            int questionCount = buf.readUnsignedShort();
            int answerCount = buf.readUnsignedShort();
            int authorityRecordCount = buf.readUnsignedShort();
            int additionalRecordCount = buf.readUnsignedShort();
            decodeQuestions(decoder, query, buf, questionCount);
            decodeRecords(decoder, query, DnsSection.ANSWER, buf, answerCount);
            decodeRecords(decoder, query, DnsSection.AUTHORITY, buf, authorityRecordCount);
            decodeRecords(decoder, query, DnsSection.ADDITIONAL, buf, additionalRecordCount);
            success = true;
            return query;
        } finally {
            if (!success) {
                query.release();
            }
        }
    }

    private static DnsQuery newQuery(ByteBuf buf, DnsQuerySupplier supplier) {
        int id = buf.readUnsignedShort();
        int flags = buf.readUnsignedShort();
        if (flags >> 15 == 1) {
            throw new CorruptedFrameException("not a query");
        }

        DnsQuery query = supplier.get(id, DnsOpCode.valueOf((byte) (flags >> 11 & 0xf)));
        query.setRecursionDesired((flags >> 8 & 1) == 1);
        query.setZ(flags >> 4 & 0x7);
        return query;
    }

    public interface DnsQuerySupplier {
        DnsQuery get(int id, DnsOpCode dnsOpCode);
    }
```

Tcp invoke
```
        DnsMessageUtil.decodeDnsQuery(frame.slice(), new DnsMessageUtil.DnsQuerySupplier() {
            @Override
            public DnsQuery get(int id, DnsOpCode dnsOpCode) {
                return new DefaultDnsQuery(id, dnsOpCode);
            }
        }, decoder);
```
		
Udp invoke 
```
        DnsMessageUtil.decodeDnsQuery(buf, new DnsMessageUtil.DnsQuerySupplier() {
            @Override
            public DnsQuery get(int id, DnsOpCode dnsOpCode) {
                return new DatagramDnsQuery(packet.sender(), packet.recipient(), id, dnsOpCode);
            }
        }, recordDecoder);
```

Reply:
Yep

Reply:
Extracted common code blocks to `DnsMessageUtil`.  :D